snapshot.go

```diff
@@ -592,6 +514,165 @@ func (snap *sqliteSnapshot) prepareWrite() error {
+			break
+		}
+
+		uniqueVersions[snapshotNode.Version] = struct{}{}
+		node := &Node{
+			key:           snapshotNode.Key,
+			subtreeHeight: snapshotNode.Height,
+			nodeKey:       NewNodeKey(snapshotNode.Version, uint32(snap.ordinal)),
+		}
+
+		stackSize := len(stack)
+		if node.isLeaf() {
+			node.value = snapshotNode.Value
+			node.size = 1
+			node._hash(snapshotNode.Version)
+			if !isStoreLeafValues {
+				node.value = nil
+			}
+
+			count++
+			if err := snap.writeSnapNode(node, snapshotNode.Version, count); err != nil {
+				return nil, nil, err
+			}
+		} else if stackSize >= 2 && stack[stackSize-1].subtreeHeight < node.subtreeHeight && stack[stackSize-2].subtreeHeight < node.subtreeHeight {
+			node.leftNode = stack[stackSize-2]
+			node.leftNodeKey = node.leftNode.nodeKey
+			node.rightNode = stack[stackSize-1]
+			node.rightNodeKey = node.rightNode.nodeKey
+			node.size = node.leftNode.size + node.rightNode.size
+			node._hash(snapshotNode.Version)
+			stack = stack[:stackSize-2]
+
+			node.leftNode = nil
+			node.rightNode = nil
+
+			count++
+			if err := snap.writeSnapNode(node, snapshotNode.Version, count); err != nil {
+				return nil, nil, err
+			}
+		}
+
+		stack = append(stack, node)
+		snap.ordinal++
+	}
+
+	if err != nil && !errors.Is(err, ErrorExportDone) {
+		return nil, nil, err
+	}
+
+	if len(stack) != 1 {
+		return nil, nil, fmt.Errorf("expected stack size 1, got %d", len(stack))
+	}
+
+	return stack[0], uniqueVersions, nil
+}
+
+func (snap *sqliteSnapshot) restorePreOrderStep(nextFn func() (*SnapshotNode, error), isStoreLeafValues bool) (*Node, map[int64]struct{}, error) {
+	var (
+		count          int
+		step           func() (*Node, error)
+		uniqueVersions = make(map[int64]struct{})
+	)
+
+	step = func() (*Node, error) {
+		snapshotNode, err := nextFn()
+		if err != nil {
+			return nil, err
+		}
+
+		node := &Node{
+			key:           snapshotNode.Key,
+			subtreeHeight: snapshotNode.Height,
+			nodeKey:       NewNodeKey(snapshotNode.Version, uint32(snap.ordinal)),
+		}
+
+		if node.isLeaf() {
+			node.value = snapshotNode.Value
+			node.size = 1
+			node._hash(snapshotNode.Version)
+			if !isStoreLeafValues {
+				node.value = nil
+			}
+		} else {
+			node.leftNode, err = step()
+			if err != nil {
+				return nil, err
+			}
+			node.leftNodeKey = node.leftNode.nodeKey
+			node.rightNode, err = step()
+			if err != nil {
+				return nil, err
+			}
+			node.rightNodeKey = node.rightNode.nodeKey
+
+			node.size = node.leftNode.size + node.rightNode.size
+			node._hash(snapshotNode.Version)
+			node.leftNode = nil
+			node.rightNode = nil
+			uniqueVersions[snapshotNode.Version] = struct{}{}
+		}
+
+		count++
+		if err := snap.writeSnapNode(node, snapshotNode.Version, count); err != nil {
+			return nil, err
+		}
+		snap.ordinal++
+
+		return node, nil
+	}
+
+	node, err := step()
+
+	return node, uniqueVersions, err
+}
+
+func (snap *sqliteSnapshot) writeSnapNode(node *Node, version int64, count int) error {
```

Comment:
I think the write logic may need to change a bit here now, if you pull `v2-alpha6` you'll see what I mean.  I needed to incorporate an additional flag to accommodate multiple write paths in snapshot.

https://github.com/cosmos/iavl/blob/640a21ec34b532d04886b6de2085c809dca6f343/snapshot.go#L30C1-L31